### PR TITLE
Asiga001track volume

### PR DESCRIPTION
**Pull request by aswritescode**

### CODE REVIEW

File: compiler/passes/analyses/track_volume.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        return
+    if (type(instruction) == Phi):
+        handle_phi(self, instruction)
+
+# We assume that the variable being created here has the minimum possible volume. This way, the compiler only proceeeds when volume is guarenteed to be correct
+def handle_phi(self, instructions : IR):
+    print("Handling phi...")
+
+    print("defs: " + str(instructions.defs))
+    print("uses: " + str(instructions.uses))
```
Comment (lilott8) [inline compiler/passes/analyses/track_volume.py]: Please use self.log.[level] for printing.

File: compiler/passes/analyses/track_volume.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+from compiler.data_structures.ir import *
+
+
+class VolumeTracker(BSAnalysis):
+
+    variable_volume = dict();                # The dict that will store the volumes of relevent variables
+
+    violation_found = False
+
+    _program = None;
```
Comment (lilott8) [inline compiler/passes/analyses/track_volume.py]: These belong in `__init__()`

File: compiler/passes/analyses/track_volume.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+                for i in program.functions[root]['blocks'][node].instructions:
+                    #print("type of i: " + str(type(i)))
+
+                    if (type(i) == Phi):
+                        print("Found a Phi node!")
+                        print(i)
+                    #print("i: " + str(i))
+                    #print("defs: " + str(i.defs))
+                    #print("uses: " + str(i.uses))
```
Comment (lilott8) [inline compiler/passes/analyses/track_volume.py]: remove rem'd lines.

File: compiler/passes/analyses/track_volume.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+            if (self.violation_found):
+                        print("Exiting...")
+                        break;
+
+        return {'name': self.name, 'result': None} # Not sure what is going on here, but its needed to prevent a crash
+
+
+
+def handle(self, instruction : IR): # The meat of the logic. Just ferries out the functions based on the type of instruction.
+    if (type(instruction) == Dispense):
```
Comment (lilott8) [inline compiler/passes/analyses/track_volume.py]: IRInstruction has an `op` field that tells you what it is.  You don't need to do `type() == ...`, you can do `instruction.op == IR.[op]`

File: test_cases/dispense_if_else.bs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+instructions:
+
+if (3<=3) {
+	a = dispense 5 units of fluid_a
+} else {
+	a = dispense 10 units of fluid_a
+}
+
+b = split a into 5
```
Comment (lilott8) [inline test_cases/dispense_if_else.bs]: All test cases belong in the `\tests\test_cases\volume\` folder.  You will then need to add a test harness to `\tests\frontend\` that handles volume tracking.  You can copy whatever of the boilerplate you need from other files.

**Review by lilott8 — CHANGES_REQUESTED** [review]
All test cases must be put in the \tests\...\... folder, whereby the pytest test harnesses can automatically run it.  All 100+ test cases should be passing, including your test cases, before another PR is issued.

File: test_cases/dispense_if_else.bs (hunk 5 above)
Comment (unleavenedbread) [reply test_cases/dispense_if_else.bs]: How do I access the global variables stored in my Pass's class? I don't see an obvious way of grabbing it using the existing functions in front_end_base.py

File: compiler/passes/analyses/track_volume.py (hunk 4 above)
Comment (unleavenedbread) [reply compiler/passes/analyses/track_volume.py]: Using Type() is actually very fast and functions well considering all Python variables are self-aware of their typing.

File: compiler/passes/analyses/track_volume.py (hunk 2 above)
Comment (unleavenedbread) [reply compiler/passes/analyses/track_volume.py]: If I move the variable declarations to within init() then I can't access them later on in my code.

Comment (tysolov) [reply compiler/passes/analyses/track_volume.py]: If they are not bound to self, then that is true; however, if you declare them, for example, as "self.variable_volume = dict()" and so on, then they are bound to the object instance.

File: compiler/passes/analyses/track_volume.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+            if (self.violation_found):
+                        #print("Exiting...")
+                        break;
+
+        return {'name': self.name, 'result': [self.violation_found, self.variable_volume]} # Returns the relevent results
+
+
+
+def handle(self, instruction : IR): # The meat of the logic. Just ferries out the functions based on the type of instruction.
```
Comment (tysolov) [inline compiler/passes/analyses/track_volume.py]: This and the following functions should be declared within the class.

You'll need to call each method as `self.<method>`, and remove the `self` parameter from the call, as the calling class instance implicitly sends itself as the first parameter.

File: compiler/semantics/ir_visitor.py [hunk 7]
```diff
@@ -755,10 +755,22 @@ def visitDispense(self, ctx: BSParser.DispenseContext):
...
             size = deff['index']
 
+        # Grab the declared volume of the variable and store it
+        _volume = 0;
+        if (type(ctx.unitTracker()) != BSParser.UnitTrackerContext):
+            _volume = 10; # Default to 10 if the volume hasn't been explicitly declared
+        else:
+            _volume = int(ctx.unitTracker().INTEGER_LITERAL().__str__())
+
+        #print("Volume: " + str(_volume)) #DEBUG
```
Comment (tysolov) [inline compiler/semantics/ir_visitor.py]: Remove dead code throughout.  For debugging, use the following pattern:
```
if self.config.debug:
   self.log.debug(<output>)
```

File: tests/frontend/test_volume.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    def test_bad_num(self, get_visitor):
+        file = "test_cases/volume/split_bad_num.bs"
+
+        tree = get_visitor(file)
+
+        vol = self.get_volume(tree, file)
+
+        assert vol[0] == True
+
+    def test_if_else(self, get_visitor):
```
Comment (tysolov) [inline tests/frontend/test_volume.py]: The result here does not seem correct.  There are two potential results that could come from the analysis on `dispense_if_else.bs`:

(1) if your analysis is capable of determining flow statically (as in this case 3<=3 is always true), then vol(a) = {5} prior to splitting, and null after the split, and each element of b should have b(i) = {1}.

(2) if your analysis ignores flow (minding not all conditionals can be statically determined, so this should be implemented either way when we have a variable conditional), then vol(a) = {5, 10} prior to split and null after split, and each element of b should have b(i) = {1, 2}.

**Review by tysolov — CHANGES_REQUESTED** [review]
Address unresolved comments prior to next review request

File: compiler/semantics/ir_visitor.py [hunk 9]
```diff
@@ -757,17 +817,18 @@ def visitDispense(self, ctx: BSParser.DispenseContext):
...
-            _volume = int(ctx.unitTracker().INTEGER_LITERAL().__str__())
-
-        #print("Volume: " + str(_volume)) #DEBUG
+            _volume = deepcopy(int(ctx.unitTracker().INTEGER_LITERAL().__str__()))
 
         # We don't have to check here, because this is a dispense.
 
-        #print("Scope_Stack: " + str(self.scope_stack[-1]))
+        if deff['name'] in self.symbol_table.get_local(deff['name'], self.scope_stack[-1]).volumes:
+            self.symbol_table.get_local(deff['name'], self.scope_stack[-1]).volumes[deff['name']].append(_volume)
```
Comment (tysolov) [inline compiler/semantics/ir_visitor.py]: Here you're using the variable name as the key to the volumes dictionary, which requires you to use some hacky manipulation to retrieve values in the volume tracker -- the whole point of using a dictionary is for uniqueness.  This works _for tested cases_, but as you are using the index of the list to retrieve the value based on the _assumed name_ from SSA-renaming, this does not generalize.  In short, the dict key should be the instruction, not the name, and when you retrieve a value in volume tracker you'll need that same line number.

You can get the _unique_ instruction id (`iid`) from the ir that is generated a few lines after this.

You will want to do the same for the mix value you parse here (unique dict items keyed by the `iid`), as there is nothing stopping us from doing:

```
a = dispense ....
a = mix x with y
```
Which would overwrite the volume attribute in the symbol table currently.

File: compiler/passes/analyses/track_volume.py [hunk 10]
```diff
@@ -26,190 +26,201 @@ def analyze(self, program: Program) -> dict: # The main function of the class
...
-    # Initialize the structures we will use later on
-    entry = dict() # The dict that will hold our new entry in the variable_volume ds
-    volumes = list() # The list that will holds the volumes stored at each index
+    def handle_dispense(self, instructions: IR):
+        name_length = len(instructions.defs['name'])
+        name_no_num = instructions.defs['name'][0 : name_length -1]
+        name_num_only = instructions.defs['name'][-1]
 
-    volumes.append(quantity) # Add the dispense'd quantity toi the volumes list. In this case we assume that size is always 1, so we only do it a single time.
+        quantity = self._program.symbol_table.get_local(instructions.defs['name'], "main").volumes[name_no_num][int(name_num_only) - 1]
```
Comment (tysolov) [inline compiler/passes/analyses/track_volume.py]: The comment I left in `ir_visitor` refers to this hacky access.  Please update accordingly using `iid` or some other unique identifier.

File: compiler/passes/analyses/track_volume.py [hunk 11]
```diff
@@ -26,190 +26,201 @@ def analyze(self, program: Program) -> dict: # The main function of the class
...
+            self.variable_volume[instructions['name']]['volumes'] = [
+                -1]  # Since volumes is a list, we wrap our single volume data in its own list. This is to avoid any issues when reading a disposed variable's entry down the line
+            self.variable_volume[instructions['name']][
+                'size'] = 0  # A disposed variable doesn't have a presence on the board. It's size is therefore zero.
 
-    else:
-        self.variable_volume[instructions['name']]['volumes'] = [-1]; # Since volumes is a list, we wrap our single volume data in its own list. This is to avoid any issues when reading a disposed variable's entry down the line
-        self.variable_volume[instructions['name']]['size']    = 0;    # A disposed variable doesn't have a presence on the board. It's size is therefore zero.
+    def handle_mix(self, instructions: IR):
+        encoded_quantity = self._program.symbol_table.get_local(instructions.defs['name'], "main").value.volume['quantity'] # The quantity stored in an encoded float value.
```
Comment (tysolov) [inline compiler/passes/analyses/track_volume.py]: This uses the `value` attribute of a variable.  You should instead use your new `volumes` attribute, as `value` is updated to the _last parsed value_ with this variable name.

Also, consider the statement: `a = mix b with c`.

You do not know if `b` and `c` have only a single possible value, as they may be defined over various control paths; hence, you should have `volumes_b` and `volumes_c`, and iterate over the combination of each pair to form the set of possible volumes of `a`.

This logic should extend to split as well.

**Review by tysolov — CHANGES_REQUESTED** [review]
Please review updated comments and address accordingly.

File: compiler/passes/analyses/track_volume.py (hunk 11 above)
Comment (unleavenedbread) [reply compiler/passes/analyses/track_volume.py]: I believe we discussed this and agreed that we could naively assume that a variable's volume is always equal to its lowest possible value.

**Review by tysolov — APPROVED** [review]

**Review by tysolov — APPROVED** [review]

**Review by lilott8 — APPROVED** [review]

**Pull request merged**